### ISSUE TITLE
mybinder.org for IPython notebooks?

### ISSUE THREAD

**Issue by norvig** [aimacode/aima-python#428]
We should look at http://mybinder.org/ and see if it is appropriate for running the Ipython/Jupyter notebooks.

**Comment by MrDupin** [comment]
I don't know if it's just me, but when I try to open any Binder project, I get stuck on a loading loop (even for small projects).

**Comment by lucasmoura** [comment]
Same problem here. At first I thought that maybe that were some problems on our notebooks to be displayed in the site. But the build proces in mybinder seems to be okay.

**Comment by MrDupin** [comment]
I researched this a bit and it seems that Binder is not very stable. A lot of [issues](https://github.com/binder-project/binder/issues) on their GitHub are about launch failing, while [parts](http://mybinder.org/feed) of their website are currently down.

Also, there doesn't seem to be much support for the platform any more. Most of their repositories haven't been updated for over a year. ([binder-build](https://github.com/binder-project/binder-build), last commit over a year ago - [repo with most recent commit](https://github.com/binder-project/binder-deploy-kubernetes))

**Comment by norvig** [comment]
OK, thanks for the feedback. We will wait until there is a more stable option available.

**Comment by reachtarunhere** [comment]
@norvig  I think it is possible to replicate their setup on a separate cloud provider. I will give it a shot with Google App Engine and get back. I feel that a dedicated server for AIMA will keep things more stable and result in better uptime. Though I still have doubts about how much we can support with the free tier service.

**Comment by sofmonk** [comment]
Apart from what @reachtarunhere mentioned, we could use [thebe](https://oreillymedia.github.io/thebe/), [thebe and jupyter](https://www.kimstebel.com/posts/thebe). thebe is a plugin to jupyter notebooks to provide interactive coding. We can possibly build more over it.

[sharecell](https://lambdaops.com/share-cell/) and [gist exec](https://github.com/rgbkrk/gistexec) [(gist exec example)](https://lambdaops.com/gistexec/?gistID=ae87a94f4ef0793d5d52) are other prototypes on which we can build more to support the needs of our project.

We may use [nbviewer](https://nbviewer.jupyter.org/) to present static notebook collection of all the chapters which may be tied to the AIMA website. Apart from those mybinder seems the option for dynamic notebooks.

@norvig Sir, if you aim running code in real time (for classroom lectures, etc) [Hydrogen](https://atom.io/packages/hydrogen) is a good tool.

@MrDupin isn't [this](https://github.com/binder-project) the binder-project it seems to be updated atleast 2 months ago?

**Comment by MrDupin** [comment]
@sofmonk, yes that is the binder-project. But the recent changes made to core components are not only few and far apart, but also miscellaneous (for example, line fixes, or updating version numbers). Unless it is supported elsewhere, the developers seem to have moved on.

**Comment by sofmonk** [comment]
mybinder is working now, maybe we should check it?

**Comment by MrDupin** [comment]
It worked for me too, but it took a long time to load (~ 2 minutes). After it got started though it worked smoothly. It seems though that the launcher is inconsistent (one time working then failing), so I'm still not 100% convinced.

**Comment by darius** [comment]
I just tried to launch it a couple times from the launch button on our README, and it failed both times with an unhelpful message like 'failed'. (I have poor connectivity here, though.)

**Comment by reachtarunhere** [comment]
@darius it often fails when you have bad connectivity given a constant link has to be established with the server for the interactivity. But even on good connections Binder fails a lot of times. This mostly has to do with the excess load on their servers (its non-commercial and free). 

One solution is hosting our own instance of [JupyterHub](https://jupyterhub.readthedocs.io/en/latest/) on Google App Engine/Digital Ocean etc. I haven't still worked out how to integrate it with GitHub auth (The other option is creating accounts manually which isn't useful in our case). Unfortunately I am finding it hard to make time to look into this. It would be great if someone can lend a hand at this.

**Comment by darius** [comment]
I'm going to see if I can run Jupyter locally first. I'm sort of busy, too...

**Comment by norvig** [comment]
I can't get the right requirements.txt for mybinder.

It is complaining that there is no compatible version of PIL.

Anyone able to help?

**Comment by ad71** [comment]
@norvig Try adding `pillow` and `Image` to requirements.txt

**Comment by dmeoli** [comment]
The README now uses the current mybinder.org/v2/gh/... links (fixed in PR #1335), so notebooks run on Binder. Closing as resolved.
